test/test_basic_logic.py

```diff
@@ -113,6 +113,57 @@ def test_sending_data(self):
             c.data_to_send() == b'\x00\x00\t\x00\x00\x00\x00\x00\x01some data'
         )
 
+    def test_sending_data_with_padding(self):
+        """
+        Single data frames with padding are encoded correctly.
+        """
+        c = h2.connection.H2Connection()
+        c.initiate_connection()
+        c.send_headers(1, self.example_request_headers)
+
+        # Clear the data, then send some data.
+        c.clear_outbound_data_buffer()
+        events = c.send_data(1, b'some data', pad_length=5)
+        assert not events
+        assert c.data_to_send() == (
+            b'\x00\x00\x0f\x00\x08\x00\x00\x00\x01'
+            b'\x05some data\x00\x00\x00\x00\x00'
+        )
+
+    def test_sending_data_with_zero_length_padding(self):
+        """
+        Single data frames with zero-length padding are encoded
+        correctly.
+        """
+        c = h2.connection.H2Connection()
+        c.initiate_connection()
+        c.send_headers(1, self.example_request_headers)
+
+        # Clear the data, then send some data.
+        c.clear_outbound_data_buffer()
+        events = c.send_data(1, b'some data', pad_length=0)
+        assert not events
+        assert c.data_to_send() == (
+            b'\x00\x00\x0a\x00\x08\x00\x00\x00\x01'
+            b'\x00some data'
+        )
+
+    def test_sending_data_with_invalid_padding_length(self):
+        """
+        Pad_lengths < 0 or > 255 are be checked for and rejected.
+        """
+        c = h2.connection.H2Connection()
+        c.initiate_connection()
+        c.send_headers(1, self.example_request_headers)
+
+        c.clear_outbound_data_buffer()
+        # 255 is valid
+        c.send_data(1, b'some data', pad_length=255)
+        with pytest.raises(ValueError):
+            c.send_data(1, b'some data', pad_length=-1)
+        with pytest.raises(ValueError):
+            c.send_data(1, b'some data', pad_length=256)
```

Comment:
We’ll want some extra tests here that we throw a `TypeError` if passing a non-integer padding length.